graph/graph.go

```diff
@@ -428,10 +427,125 @@ func (graph *Graph) Heads() (map[string]*image.Image, error) {
+	// Store the layer. If layerData is not nil, unpack it into the new layer
+	if layerData != nil {
+		if img.Size, err = graph.driver.ApplyDiff(img.ID, img.Parent, layerData); err != nil {
+			return err
+		}
+	}
+
+	if err := graph.saveSize(root, int(img.Size)); err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(jsonPath(root), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	return json.NewEncoder(f).Encode(img)
+}
+
+// loadImage fetches the image with the given id from the graph.
+func (graph *Graph) loadImage(id string) (*image.Image, error) {
+	root := graph.imageRoot(id)
+
+	// Open the JSON file to decode by streaming
+	jsonSource, err := os.Open(jsonPath(root))
+	if err != nil {
+		return nil, err
+	}
+	defer jsonSource.Close()
+
+	img := &image.Image{}
+	dec := json.NewDecoder(jsonSource)
+
+	// Decode the JSON data
+	if err := dec.Decode(img); err != nil {
+		return nil, err
+	}
+	if err := image.ValidateID(img.ID); err != nil {
+		return nil, err
+	}
+
+	if buf, err := ioutil.ReadFile(filepath.Join(root, "layersize")); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		// If the layersize file does not exist then set the size to a negative number
+		// because a layer size of 0 (zero) is valid
+		img.Size = -1
+	} else {
+		// Using Atoi here instead would temporarily convert the size to a machine
+		// dependent integer type, which causes images larger than 2^31 bytes to
+		// display negative sizes on 32-bit machines:
+		size, err := strconv.ParseInt(string(buf), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		img.Size = int64(size)
+	}
+
+	return img, nil
+}
+
+// saveSize stores the `size` in the provided graph `img` directory `root`.
+func (graph *Graph) saveSize(root string, size int) error {
+	if err := ioutil.WriteFile(filepath.Join(root, "layersize"), []byte(strconv.Itoa(size)), 0600); err != nil {
+		return fmt.Errorf("Error storing image size in %s/layersize: %s", root, err)
+	}
+	return nil
+}
+
+// SetCheckSum sets the checksum for the image layer to the provided value.
+func (graph *Graph) SetCheckSum(id, checksum string) error {
+	root := graph.imageRoot(id)
+	if err := ioutil.WriteFile(filepath.Join(root, "checksum"), []byte(checksum), 0600); err != nil {
+		return fmt.Errorf("Error storing checksum in %s/checksum: %s", root, err)
+	}
+	return nil
+}
+
+// GetCheckSum gets the checksum for the provide image layer id.
+func (graph *Graph) GetCheckSum(id string) (string, error) {
+	root := graph.imageRoot(id)
+	cs, err := ioutil.ReadFile(filepath.Join(root, "checksum"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return "", nil
+		}
+		return "", err
+	}
+	return string(cs), err
+}
+
+// RawJson returns the JSON representation for an image as a byte array.
+func (graph *Graph) RawJson(img *image.Image) ([]byte, error) {
```

Comment:
This function is strange since it only uses the image ID and pulls the contents directly from disk. The best move is probably to just take in the `id` instead of the entire `*image.Image`